Added tests for parameter examples

esteve opened the pull request:
Added tests for parameter examples

Connects to ros2/ros2#143

dirk-thomas commented:
The test should not only cover the return code of the executable but also check for the expected output.

esteve commented:
The tests pass locally on Linux, here are the CI jobs:

http://ci.ros2.org/job/ros2_batch_ci_linux/574/
http://ci.ros2.org/job/ros2_batch_ci_osx/457/
http://ci.ros2.org/job/ros2_batch_ci_windows/616/

dirk-thomas commented on rclcpp_examples/test/test_single_executable_example.py.in:
```diff
@@ -0,0 +1,81 @@
+        elif line.startswith(b'rc'):
+            continue
+        else:
+            output_lines.append(line + b'\r\n')
```
`\r\n`?

esteve replied:
So to make it compatible with Windows.

dirk-thomas replied:
How does that work in non-Windows then?

esteve replied:
The fixtures use `\r\n` as line separator and the code here ensure that the output from an executable is stored with `\r\n` too.

dirk-thomas replied:
Instead of relying on Windows line endings in the fixture I would propose to remove the newlines altogether from the check. Simply split the content using `splitlines()` and perform the comparison on the lists.

wjwwood replied:
If you need to represent a new line that is portable you can also use `os.linesep`, but avoiding it would be better if that's possible.

dirk-thomas commented on rclcpp_examples/test/test_single_executable_example.py.in:
```diff
@@ -0,0 +1,81 @@
+            continue
+        elif line.startswith(b'Description'):
+            continue
+        elif line.strip() == b'':
```
`elif not line.strip():`

esteve replied:
I prefer my version as it is more explicit.

dirk-thomas commented on rclcpp_examples/test/test_single_executable_example.py.in:
```diff
@@ -0,0 +1,81 @@
+        elif line.startswith(b'Internals'):
+            continue
+        elif line.startswith(b'rc'):
+            continue
```
What are all these skipped lines?

esteve replied:
Depending on the configuration OpenSplice may log its output to stdout.

dirk-thomas replied:
Should the tests then just override the logging env variable? Then it doesn't need to deal with all of this. Also in a future version OpenSplice might just add / change some of that.

esteve replied:
Sounds good, how do I do that?

dirk-thomas replied:
By setting the env vars (https://github.com/ros2/rmw_opensplice/blob/3fd77209d3d905c22ae03f9077bbc196a63b6271/opensplice_cmake_module/env_hook/opensplice.sh.in#L42-L43) to different values.

esteve replied:
I've set `OSPL_INFOFILE` and `OSPL_ERRORFILE` to `os.devnull` like this:

```python
if __name__ == '__main__':
    os.environ['OSPL_INFOFILE'] = os.devnull
    os.environ['OSPL_ERRORFILE'] = os.devnull
    test_executable()
```

and OpenSplice still logs its output to the console. Should this code go elsewhere? Thanks.

dirk-thomas replied:
Nosetests does not invoke the script with `__name__ == '__main__'`. It invokes the `test*` functions directly.

esteve replied:
Done. I've set verbosity level to 8 (no output) and the tests pass.

dirk-thomas commented on rclcpp_examples/test/test_single_executable_example.py.in:
```diff
@@ -0,0 +1,81 @@
+from launch import LaunchDescriptor
+from launch.exit_handler import primary_exit_handler
+from launch.launcher import DefaultLauncher
+from launch.output_handler import LineOutput, FileOutput
```
One import per line.

esteve replied:
Fixed.

dirk-thomas commented on rclcpp_examples/test/test_single_executable_example.py.in:
```diff
@@ -0,0 +1,81 @@
+
+    output_lines = []
+
+    expected_output = open('@RCLCPP_EXAMPLES_EXPECTED_OUTPUT@', 'rb').read()
```
Please make sure to close the file handle.

esteve replied:
Fixed.

esteve commented:
I've changed the test to test lines (and thus skipping \n or \r\n) and squashed the commits. New CI jobs:

http://ci.ros2.org/job/ros2_batch_ci_linux/577/
http://ci.ros2.org/job/ros2_batch_ci_osx/461/
http://ci.ros2.org/job/ros2_batch_ci_windows/617/

esteve commented:
Updated code to ignore lines from RTI Connext:

http://ci.ros2.org/job/ros2_batch_ci_linux/580/
http://ci.ros2.org/job/ros2_batch_ci_osx/464/

Waiting for Windows slave to be enabled again.

esteve commented:
OSX and Windows (OpenSplice, RTI Connext static) jobs:

http://ci.ros2.org/job/ros2_batch_ci_osx/465/
http://ci.ros2.org/job/ros2_batch_ci_windows/622/
http://ci.ros2.org/job/ros2_batch_ci_windows/623/

esteve commented:
The tests for the parameters examples pass on OSX and Windows. Windows fails on an unrelated test. Here's the CI job for Linux:

http://ci.ros2.org/job/ros2_batch_ci_linux/581/

dirk-thomas commented on rclcpp_examples/test/test_single_executable_example.py.in:
```diff
@@ -0,0 +1,72 @@
+
+
+def setup():
+    os.environ['OSPL_VERBOSITY'] = '8 ' # 8 = OS_NONE
```
Is the white space after the number necessary?

dirk-thomas replied:
PEP8

esteve replied:
Fixed.

dirk-thomas commented on rclcpp_examples/test/test_single_executable_example.py.in:
```diff
@@ -0,0 +1,72 @@
+
+def setup():
+    os.environ['OSPL_VERBOSITY'] = '8 ' # 8 = OS_NONE
+
```
PEP8

esteve replied:
Fixed.

dirk-thomas commented on rclcpp_examples/test/test_single_executable_example.py.in:
```diff
@@ -0,0 +1,72 @@
+    expected_output_lines = None
+
+    with open('@RCLCPP_EXAMPLES_EXPECTED_OUTPUT@', 'rb') as f:
+      expected_output_lines = [line.strip() for line in f.read().splitlines()]
```
PEP8

dirk-thomas replied:
Why `strip`?

esteve replied:
PEP8 fixed and `strip` removed.

dirk-thomas commented on rclcpp_examples/test/test_single_executable_example.py.in:
```diff
@@ -0,0 +1,72 @@
+            continue
+        elif line.startswith(b'Expires on'):
+            continue
+        output_lines.append(line.strip())
```
Why `strip`?

esteve replied:
Removed.

esteve commented:
New CI jobs:

http://ci.ros2.org/job/ros2_batch_ci_linux/583/
http://ci.ros2.org/job/ros2_batch_ci_osx/467/
http://ci.ros2.org/job/ros2_batch_ci_windows/627/ (OpenSplice)

dirk-thomas commented on rclcpp_examples/test/test_single_executable_example.py.in:
```diff
@@ -0,0 +1,74 @@
+    expected_output_lines = None
+
+    with open('@RCLCPP_EXAMPLES_EXPECTED_OUTPUT@', 'rb') as f:
+        expected_output_lines = [line for line in f.read().splitlines()]
```
Why still use a list comprehension here?

esteve replied:
True, not needed anymore.

dirk-thomas commented on rclcpp_examples/test/test_single_executable_example.py.in:
```diff
@@ -0,0 +1,74 @@
+    for line in output_handler.stdout_data.getvalue().splitlines():
+        if line.startswith(b'pid'):
+            continue
+        elif line.startswith(b'rc'):
```
Better `if` instead of `elif` since the if condition block always continues.

Same below.

esteve replied:
Done.

dirk-thomas replied:
Pushed?

esteve replied:
This time yes :-)

esteve commented:
CI jobs cancelled and retriggered:

http://ci.ros2.org/job/ros2_batch_ci_linux/584/
http://ci.ros2.org/job/ros2_batch_ci_osx/468/
http://ci.ros2.org/job/ros2_batch_ci_windows/627/ (OpenSplice)

esteve commented:
The tests for the examples pass, this should be ready for merging.

dirk-thomas commented on rclcpp_examples/test/test_single_executable_example.py.in:
```diff
@@ -0,0 +1,74 @@
+        if line.startswith(b'pid'):
+            continue
+        if line.startswith(b'rc'):
+          continue
```
Indentation.

Same below.

Please run `ament_pep8` on this template.

gerkey commented:
I branched from here to add support for running multiple processes in the `test_talker_listener` branch.

dirk-thomas commented on rclcpp_examples/test/test_single_executable_example.py.in:
```diff
@@ -0,0 +1,75 @@
+    ld.add_process(
+        cmd=['@RCLCPP_EXAMPLES_EXECUTABLE@', 'test_executable'],
+        name='test_executable',
+        exit_handler=primary_exit_handler,
```
For a single process the `primary_exit_handler` doesn't make any difference. But when adding a second process they can't both be primary. Therefore I would recommend to use the default exit handler instead.

esteve replied:
Fixed.

dirk-thomas commented:
lgtm
